### DISCUSSION TITLE
Session Timeout Questions

### DISCUSSION THREAD

**Discussion by taejinseo** [keycloak/keycloak#22920] [Q&A]
I have a question about "Session Timeout per Realm" and "Session Timeout per Client" in Keycloak.

I thought the policy for Session Timeout is that the "Realm" criteria is applied first, and if there is a "Client" Session Timeout setting, the client's setting is applied first.

However, the actual behavior is that the "Realm" setting is always applied first and the token is regenerated when the session timeout occurs.

 Is there any other option to make the session timeout behavior order "Client" first, and then "Realm" settings only if there are no "Client" settings? 
This does not seem to be a specific version issue.

We have tried "15.0.2" and "22.0.0" and both versions are the same.

We are looking for help to resolve this.

**Comment by justin-tay** [comment]
It's not really clear what you mean by "Session Timeout per Realm" and "Session Timeout per Client" and what you are expecting to happen after setting those values.

For instance in Keycloak 22 if I configure this

![realm](https://github.com/keycloak/keycloak/assets/49700559/4136faf7-7bfa-4384-88c2-6f7d0b46b284)

Then the `refresh_expires_in` in the token endpoint is `1200`.

If I then configure

![client](https://github.com/keycloak/keycloak/assets/49700559/dbfe20de-896d-4b02-8a66-82daec8b5e56)

Then the `refresh_expires_in` in the token endpoint is `1500`.

**Reply by taejinseo** [reply]
Thank you for your answer.
Here is what we want to implement

--- 

- Realm Setting : session timeout - 8 Hours
- "A" Client Setting : session timeout - 4 Hours
- "B" Client Setting : session timeout - N/A

### Case1 
A Client Login -> B Client Login -> (4 Hours Later) -> A/B Logout

### Case2
B Client Login -> A Client Login -> (4 Hours Later) -> A/B not Logout -> (8 Hour Later) -> A/B Logout

---

I'm trying to find the right settings for this, but it's not easy.

**Reply by justin-tay** [reply]
In the admin console, where and what are the settings that you are modifying?

When you say client session logout, what do you mean? Just to confirm, your are referring to the client session and not the sso user session?

**Comment by min-sung-kwon** [comment]
I'm currently using Keycloak 22. Within a single Realm, I have multiple Clients.

I'd like to set session timeouts for each Client so that when the session expires, it redirects to the login page. 
However, it seems that the `SSO session timeout` setting at the Realm level is affecting this behavior, and the session keeps getting extended without redirecting to the login page.

Is there a way to make it so that the `Client session timeout` settings are not affected by the `SSO session timeout` setting, and sessions expire according to the `Client session timeout` values?
If possible, could you please guide me on how to achieve this? 

Thank you.

**Reply by justin-tay** [reply]
It sounds like you are confusing the SSO session and the client session. They are not the same thing even though they are somewhat related to each other.

The client session is generally related to the ability of the client to use a refresh token to get new tokens. When the client does this, it extends the client session (gets a new refresh token with an extended expiry) and the related SSO user session.

The SSO session is related to a Single-Sign On user. So the user signs on to the realm and then he doesn't have to individually sign on to all the applications on the realm. Keycloak stores a cookie on the user's browser to track this. Given what this does, there is only a realm configuration for this setting.

So it's not that the client session isn't expiring, they are, it's just that if you just query the authorization endpoint again Keycloak is going to automatically log him back on again if the SSO session is still valid.

It's hard to say what you want, like do you even want Single-Sign On in the first place? You could remove the `Cookie` authenticator from the `browser` flow.

If your applications want the user to authenticate again despite having a valid SSO session, you can get the application to send `prompt=login` with the authorization request. Note that your application would need to verify the `auth_time` claim is as expected.
